### ISSUE TITLE
Enum cases incorrectly generated when containing enum name

### ISSUE THREAD

**Issue by rebello95** [apple/swift-protobuf#759]
**Issue**
When an enum case's first few letters contain the name of the enum itself (even if separated by underscores), the plugin strips those characters from the enum name.

**Example**
```proto
syntax = "proto3";

message Test {
    Promos promos = 1;

    enum Promos {
        PROMO_SCREEN = 0;
        TEST_ONE = 1;
    }
}
```

**Actual**
```swift
public enum Promos: SwiftProtobuf.Enum {
	public typealias RawValue = Int
	case creen // = 0
	case testOne // = 1
	case UNRECOGNIZED(Int)
}
```

**Expected**
```swift
public enum Promos: SwiftProtobuf.Enum {
	public typealias RawValue = Int
	case promoScreen // = 0
	case testOne // = 1
	case UNRECOGNIZED(Int)
}
```

**Reproducibility**
100% reproducible on SwiftProtobuf 1.0.3

**Comment by tbkka** [comment]
If valid proto files generate invalid Swift code, that is certainly a bug.  Please provide more details so we can fix it.  (As you observed, your example does not demonstrate such a bug.)

As the error message points out, C# protobuf also strips enum names in a similar fashion, so there is ample precedent for this behavior.

We certainly cannot change this soon.   It would break existing code for people already using SwiftProtobuf.  We can only make such breaking changes at a major version, and we do not at this time have other changes that would justify creating a 2.0 release.

Note that I am happy to discuss such a change.  If there's a strong consensus that we should change it, then we can certainly put this into a list of potential changes for a future 2.0 release.

**Comment by thomasvl** [comment]
To provide some more history –

Top level enums in C (C++ generation), don't get scoped to the enum name, so the common practice is to repeat the enum name on all the values.  Lots of `.proto` files do this.  But for the languages that do scope the enum values to the names, the prefix makes for redundant names.

Since Apple also did this sorta prefix stripping for the System Frameworks with the updated Swift Naming guidelines in the 3.x/4.0 time frame, it seemed to make sense to try to model this on the proto side also to get the better scoping Swift provides.

The one you call out _PROMO_SCREEN_ is a bug in the current impl that should get fixed.  Hopefully I'll get a chance to look at that in the near future is someone else doesn't beat me to it.

**Comment by thomasvl** [comment]
Finally got a chance to look back at the code this morning.  The logic in the generator matches what the C# ones does, and it also have the same weakness.  Both work by stripping the underscores characters and then just case insensitively matching.

We could retool things to use the underscores and or changes in case to indicate a _word_ in the names, then match words.  The one gap that jumps out at me is an enum of _foo_, but a values of _foobar_ and _foobaz_ would no longer strip down to just _bar_ and _baz_.  But maybe some of those things can be resolved by checking if the value was mixed case at all, and what the case of the first character after the prefix was.

**Comment by rebello95** [comment]
Thanks for the details and context @thomasvl. The goal of prefix stripping in generated Swift code definitely makes sense.

If we're okay with making a change (which would be breaking, as @tbkka mentioned), I can take a look into it myself as well. It seems reasonable in this case to classify a "word" using both camelCase and SNAKE_CASE handling.

IMO it makes sense to avoid stripping prefixes for enums like the example you listed if they involve neither camel case or snake case:

    enum foo {
        foobar = 0; // -> foobar
        foo_baz = 1; // -> baz
        foo_foobar = 2; // Disallowed by protoc
    } 

    enum foo_bar {
        foo_bar_abc = 0; // -> abc
        foo_abc = 1; // -> foo_abc
    } 

    enum MyFoo {
        MyFooBar = 0; // -> bar
        MyFooBaz = 1; // -> baz
    } 

There are a few interesting [test cases](https://github.com/apple/swift-protobuf/blob/1.1.0/Tests/SwiftProtobufPluginLibraryTests/Test_NamingUtils.swift#L78) in the code base which would likely change as well:

      (prefix: "foo_bar", str: "foobar_bAZ", expected: "bAZ" ),
      (prefix: "_foo_bar", str: "foobar_bAZ", expected: "bAZ" ),
      (prefix: "foo__bar_", str: "_foo_bar__baz", expected: "baz"),

With the new changes, these would become (notice that the number of underscores remains irrelevant as it does today):

      (prefix: "foo_bar", str: "foobar_bAZ", expected: "foobar_bAZ" ),
      (prefix: "_foo_bar", str: "foobar_bAZ", expected: "foobar_bAZ" ),
      (prefix: "foo__bar_", str: "_foo_bar__baz", expected: "baz"),
      (prefix: "FooBar", str: "FooBarBaz", expected: "baz"),
      (prefix: "foo", str: "_foo_bar__baz", expected: "bar_baz"),

What are your thoughts on this?

**Comment by thomasvl** [comment]
> `foo_foobar = 3; // Disallowed by protoc`

What's the error on this one?  Sounds like they added some more checks since I last looked.

At first glance, it seems reasonable.  The test cases will also need to be expand for camel vs. underscore, as well as still ignoring the number of underscores:

```
  (prefix: "foo_bar", str: "FooBarBaz", expected: "Baz")
  (prefix: "FooBar", str: "foo_bar_baz", expected: "baz")
  (prefix: "foo__bar", str: "FooBarBaz", expected: "Baz")
  (prefix: "FooBar", str: "_foo__bar__baz", expected: "baz")
  ...
```

As far as breaking, we've got some other stuff we've been looking at doing that might also be breaking (I'm trying to get the cycles to work on them), so now is as good of time as any.  So if you are interested, go for it!

**Comment by rebello95** [comment]
> > `foo_foobar = 3; // Disallowed by protoc`

> What's the error on this one? Sounds like they added some more checks since I last looked.

Sorry, typo'd there - the one that fails is `foo_bar`, not `foo_foobar` in that case. The error is:
> When enum name is stripped and label is PascalCased (Bar), this value label conflicts with foobar. This will make the proto fail to compile for some languages, such as C#.

It looks like some of the test cases you listed are in conflict with the ones I posted:

    (prefix: "foo_bar", str: "foobar_bAZ", expected: "foobar_bAZ" )

versus:

    (prefix: "foo_bar", str: "FooBarBaz", expected: "Baz")

(Note that the first one does not take into account underscores). Looking over it again, I think what you're proposing (second example above) makes sense and would still solve the original problem 👍 Hoping to get a chance to look into it soon, and I'll be sure to PR if so!

**Comment by JG-NI** [comment]
We are hitting this bug in C#.  It is clearly wrong and hard to fix outside of protoc given that it works for some client languages and not others.  I think it should be fixed asap and if you are worried about recompiling existing protos, add a command line option for the old behavior. If you won't do that, add a command line option to select the correct behavior.

**Comment by JG-NI** [comment]
Yes, not using swift. Just commenting that is it amazing this hasn't been fixed in general given how wrong the output is.
